src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,759 @@
+	uint64 transactionClockValue = GetHighestClockInTransaction(transactionNodeList);
+
+	/* Persist the transactionId along with the logical commit-clock timestamp */
+	LogTransactionCommitClock(transactionId, transactionClockValue);
+
+	/* Adjust all the nodes with the new clock value */
+	AdjustClocksToTransactionHighest(transactionNodeList, transactionClockValue);
+}
+
+
+/*
+ * LogTransactionCommitClock registers the committed transaction along
+ * with the commit clock.
+ */
+static void
+LogTransactionCommitClock(char *transactionId, uint64 transactionClockValue)
+{
+	Datum values[Natts_pg_dist_commit_transaction];
+	bool isNulls[Natts_pg_dist_commit_transaction];
+	uint64 clockLogical = GET_LOGICAL(transactionClockValue);
+	uint32 clockCounter = GET_COUNTER(transactionClockValue);
+
+	ereport(DEBUG1, (errmsg("Persiting transaction %s with "
+							"clock logical (%lu) and counter(%u)",
+							transactionId, clockLogical, clockCounter)));
+
+	/* form new transaction tuple */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	uint64 transactionTimestamp = GetEpochTimeMs();
+
+	values[Anum_pg_dist_commit_transaction_transactionid - 1] =
+		CStringGetTextDatum(transactionId);
+	values[Anum_pg_dist_commit_transaction_clock_logical - 1] =
+		Int64GetDatum(clockLogical);
+	values[Anum_pg_dist_commit_transaction_clock_counter - 1] =
+		Int64GetDatum(clockCounter);
+	values[Anum_pg_dist_commit_transaction_timestamp - 1] =
+		Int64GetDatum(transactionTimestamp);
+
+	/* open transaction relation and insert new tuple */
+	Relation pgDistCommitTransaction =
+		table_open(DistCommitTransactionRelationId(), RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(pgDistCommitTransaction);
+	HeapTuple heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+
+	CatalogTupleInsert(pgDistCommitTransaction, heapTuple);
+
+	CommandCounterIncrement();
+
+	/* close relation and invalidate previous cache entry */
+	table_close(pgDistCommitTransaction, NoLock);
+}
+
+
+/*
+ * Initialize the shared memory clock value to the highest clock
+ * persisted. This will protect from any clock drifts.
+ */
+void
+InitClockAtBoot(void)
+{
+	if (creating_extension)
+	{
+		/* No catalog present yet */
+		return;
+	}
+
+	SpinLockAcquire(&logicalClockShmem->clockMutex);
+
+	/* Avoid repeated initialization */
+	if (logicalClockShmem->clockInitialized == true)
+	{
+		SpinLockRelease(&logicalClockShmem->clockMutex);
+		return;
+	}
+
+	/*
+	 * Set the flag before executing a distributed query
+	 * (else it might trigger this routine recursively.)
+	 */
+	logicalClockShmem->clockInitialized = true;
+
+	/* Start with the wall clock value */
+	uint64 epochValue = GetEpochTimeMs();
+	logicalClockShmem->clusterClockValue = START_NEW_CLOCK(epochValue);
+
+	SpinLockRelease(&logicalClockShmem->clockMutex);
+
+	/*
+	 * Version checking is a must for the below code as it accesses the
+	 * new catalog created in the new version.
```

Comment:
setting citus.enable_versions_checks to false is not something that we recommend to users and hence this is their own risk (plus, iirc we shouldn't be promoting this GUC in docs already)

for this reason we don't need to temporarily set EnableVersionChecks to true

Reply:
You are spot-on, but unfortunately one of our upgrade test in our pipeline is setting this flag, thus causing issues. I'm not comfortable putting this hack just for the sake of our single test, but I couldn't find an easy solution.

Reply:
> one of our upgrade test in our pipeline is setting this flag, causing issues. 

Couldn't quite follow; if this flag being set to true causes some sort of issues, then why we're enabling it here ? :)

Reply:
It's setting the flag to false
class **CitusUpgradeConfig**(CitusBaseClusterConfig):
    def __init__(self, arguments):
        super().__init__(arguments)
        self.pre_tar_path = arguments["--citus-pre-tar"]
        self.post_tar_path = arguments["--citus-post-tar"]
        self.temp_dir = "./tmp_citus_upgrade"
        **self.new_settings = {"citus.enable_version_checks": "false"}**

Reply:
I'm closing this thread, we have the same discussion in another thread, please see below